src/watertap_contrib/seto/costing/units/chemical_softening_zo.py

```diff
@@ -0,0 +1,968 @@
+#################################################################################
+# WaterTAP Copyright (c) 2020-2023, The Regents of the University of California,
+# through Lawrence Berkeley National Laboratory, Oak Ridge National Laboratory,
+# National Renewable Energy Laboratory, and National Energy Technology
+# Laboratory (subject to receipt of any required approvals from the U.S. Dept.
+# of Energy). All rights reserved.
+#
+# Please see the files COPYRIGHT.md and LICENSE.md for full copyright and license
+# information, respectively. These files are also available online at the URL
+# "https://github.com/watertap-org/watertap/"
+#################################################################################
+
+
+from pyomo.environ import (
+    Constraint,
+    Expression,
+    Reals,
+    NonNegativeReals,
+    log,
+    Var,
+    Param,
+    Set,
+    Suffix,
+    value,
+    check_optimal_termination,
+    units as pyunits,
+)
+
+from watertap.costing.util import register_costing_parameter_block
+from idaes.core.util.misc import StrEnum
+from ..util import (
+    make_capital_cost_var,
+    make_fixed_operating_cost_var,
+)
+
+
+def build_lime_cost_param_block(blk):
+
+    blk.cost = Param(
+        initialize=0.171,
+        units=pyunits.USD_2020 / pyunits.kg,
+        doc="Cost of CaO $/kg",
+    )
+
+    blk.purity = Param(
+        mutable=True,
+        initialize=1,
+        doc="CaO purity",
+        units=pyunits.dimensionless,
+    )
+
+    costing = blk.parent_block()
+    costing.add_defined_flow("lime", blk.cost / blk.purity)
+
+
+def build_soda_ash_cost_param_block(blk):
+
+    blk.cost = Param(
+        initialize=0.65,
+        units=pyunits.USD_2020 / pyunits.kg,
+        doc="Cost of Na2CO3 $/kg",
+    )
+
+    blk.purity = Param(
+        mutable=True,
+        initialize=1,
+        doc="Na2CO3 purity",
+        units=pyunits.dimensionless,
+    )
+
+    costing = blk.parent_block()
+    costing.add_defined_flow("soda ash", blk.cost / blk.purity)
+
+
+def build_mgcl2_cost_param_block(blk):
+
+    blk.cost = Param(
+        initialize=1.5,
+        units=pyunits.USD_2020 / pyunits.kg,
+        doc="Cost of MgCl2 $/kg",
+    )
+
+    blk.purity = Param(
+        mutable=True,
+        initialize=1,
+        doc="MgCl2 purity",
+        units=pyunits.dimensionless,
+    )
+
+    costing = blk.parent_block()
+    costing.add_defined_flow("mgcl2", blk.cost / blk.purity)
```

Comment:
In the WT costing package, we have the following: https://github.com/watertap-org/watertap/blob/main/watertap/costing/watertap_costing_package.py#L179-L185

At first I was going to suggest conforming to what we have in WaterTap, since this is meant to be part of WaterTAP. However, we may want to incorporate what you did here to WT costing package. What do you think?

Reply:
I had never seen that before and I think it is an exception to how other chemical flows/costs are handled, for example from the IX costing method: https://github.com/watertap-org/watertap/blob/main/watertap/costing/units/ion_exchange.py

I think @bknueven changed how we did this a while go so we didn't have a bunch of chemical costs on the main costing block when most of the time they are only used for specific units. So in short - this is already how it is done in WaterTAP and I don't know how ``magnesium_chloride_cost`` slipped into the WaterTAP costing pkg in this form.

Reply:
I agree. Not included chemical costs in the main costing block makes it cleaner. It is easier to track which flows are included in respective unit models and not have to go back to the main costing block and update it.